### ISSUE TITLE
Milestone and roadmap management

### ISSUE THREAD

**Issue by oliviermaridat** [kanboard/kanboard#971]
I have a proposal for a better management of milestones through task links. I already have a proof of concept in my laptop, but I would like some feedbacks before to put more effort in it.
Let me explain my objectives and my proposal.

I would like to:
- have proper milestone and roadmap features
- simplify the task link view

I would also like to enhance the time tracking system. Actually, I know there is a lot of work about this at the moment, so I won't treat this now.
### Milestone and Roadmap

My idea is to treat "is a milestone of" links differently from the others. This will provide a better "Milestone" management, and clarify the Task Link view that was already too complex but still not enough complex to cover all use cases. Therefore, in the task view, instead of having:
- Links: list of link labels and their related tasks + time tracking per task
- Subtasks: list of subtasks + time tracking per subtask
- Time tracking

We could have:
- Milestone: list of tasks that targets this milestone + time tracking per task + total time tracking
  - (optionaly) A click on a task id displays the "dropdown menu", like in the board view
  - (optionaly) Each task is displayed as a collapsed colored card, like in the board view
  - (optionaly but difficult) Each task card can be drag and dropped to change the order
- Subtasks: list of subtasks + time tracking per subtask
  - (optionaly) + total time tracking
  - (optionaly) + list of tasks that has a link "is a child of" (they can be seen as a different kind of subtask)
- Time tracking (refactoring may be required, but this is another discussion ;-D)
- Links: a more simple list of link labels and their related tasks (without "is a milestone of")
  - Maybe some columns may be removed, like "Time tracking"...

This can be rendered like this:

![milestone](https://cloud.githubusercontent.com/assets/953989/8475686/1f6ded3c-20ba-11e5-96ad-e6022b2d7e91.png)
![links](https://cloud.githubusercontent.com/assets/953989/8475786/d9050140-20ba-11e5-9993-6076f7bb3023.png)

With this solution, the view of non-Milestone task is simplified, and the view of Milestone task provides better information.

After that, a new "Roadmap View" can be added to Kanboard. This view will display the list of all Milestones (i.e. Task linked as "is a milestone of") of a board and their related tasks, as I just described for the task view.

So what do you think of all this? Good idea / bad idea? Suggestions, questions... ?

**Comment by fguillot** [comment]
Instead of using task links, why not build something just for managing milestones?

I look at several software and I like the simplicity of Github issue. You can create a milestone with a name, description and a due date. Any issue can be part of a milestone.

If we apply the same thing to Kanboard, milestones can be a new section, tasks associated to a milestone can come from any project. That allow us to display milestones differently on the calendar or in a probable future Gantt diagram. All milestones can be displayed easily on a single page so you got a roadmap view.

Just an idea.

**Comment by oliviermaridat** [comment]
Thank you for your response @fguillot. Yep, Github's issue milestone system is indeed very nice and simple. I will be glad to see something that simple to use in Kanboard :sparkles:

I would prefer to use the existing Kanboard data model instead of adding a new section for milestone. A task already have title, description and due date like Github's milestone, and it is already possible to link tasks together, especially with a "is milestone" link. In my opinion, adding a new "milestone" section will add too much complexity and increase the code size uselessly.

I like the idea of adding a new page "Add a milestone" to create a new task with only title, description and due date (other fields are hidden). And also a link to add easily a task to an existing milestone. And of course, I would really like to have a "roadmap" view, and to be able to see milestones in the "calendar" view (and I am glad to hear about the "Gantt" view ;-D). This is possible with a little more work.
Milestones can already be recognised from simple tasks thanks to their "is a milestone of" link, however we can simplify this by adding a new field "is_milestone" to task. As a bonus, this allows to display a quick list of available milestones when adding a task to a milestone (like in Github).

Is it ok if I improve my PR with these proposals?

**Comment by BlueTeck** [comment]
Subtask could be in the same table to (it would be very easy to implement #873 #943 #831 and others). 
Instead of adding is_milestone i would prefer a "type" field it is more flexible (like the wordpress post table https://codex.wordpress.org/Post_Types).

**Comment by fguillot** [comment]
For me, managing milestones with links seems to be a workaround because you apply a different behavior to link marked as "milestone".

People are able to rename the link labels or remove completely this kind of links from the settings page, and don't forget that link label are translated. So if someone remove all default link labels and create new ones, you got unexpected behavior because you assume that the id 9 is a milestone: https://github.com/fguillot/kanboard/pull/1009/files#diff-a804d0f2e444a8643d7763ccdae3dc9dR36 The end-user will not necessary understand why there is a different behavior for that kind of links.

Using a different data model for milestones can offer more flexibility in the future while remaining simple. Because a milestone is almost the same thing as a sprint, so you can easily attach some stats to your milestone/sprint such as the velocity or burndown chart. You can also display a progress percentage like Gitlab milestones etc. Basically, it's just a new table with 3 or 4 columns and new column "milestone_id" in the table tasks. The order of tasks can be the same order as in the board.

For the Gantt chart, add a new column "estimate_start_date" in the task table and reuse the "estimated_time" field, order the task by column/position and that's it. For the ui, I found this jquery plugin that can be easily improved https://github.com/thegrubbsian/jquery.ganttView

So let me think about that...

**Comment by oliviermaridat** [comment]
Clearly, the actual spirit of Kanboard is in line with what you said @fguillot (which is logical, because you are the spirit of Kanboard): prefer a new data model  than specialising an existing one.

In am more in line with the proposal of @BlueTeck. For me "milestone" and "subtask" are specialised tasks and we can even add other specialised types in the future. This vision implies to also add specialised links (I did not want to spent too much time, so I used link 9 "is a milestone of" in my PR, but this is indeed no enough). I do not have enough comparison to say what is the best approach.

Anyway, I like your ideas about milestone, progress bar, Gantt ... !

**Comment by oliviermaridat** [comment]
I have a new idea to provide specific behaviours to a link: adding options to links. 2 or 3 options only may be required:
- starts after
- ends before
- "special list"

### starts after
For example "starts after". If the link "is a milestone of / targets milestone" has the option "starts after", all tasks linked with "targets milestone" to another task called M will have a "started_date" after the "started_date" of task M. The best is to add a new "started_date_computed" field to task (or something similar) that can be displayed in addition to "started_date" in calendar and Gantt views. 

### ends before
"ends before" will have the same behaviour but with "due_date".

### ends before
"special list" (yes, this is not very explicit, better name required) will allows to implement the "Milestone" view presented in the first message of this discussion. A link with the "special list" option will be removed from the "Task links" view, but will have its own way of listing its linked tasks.

Conclusion: "is a milestone of" will have the options "starts after", "ends before" and "special list"; "is blocked by" will have the option "starts after" only.

We may imagine to add new options, for example to "not display in boards" (for "is a parent of" link).

**Comment by fguillot** [comment]
Closing this issue, I think you made your own plugin.
